wallet: optimize migration process, batch db transactions

furszy opened the pull request:
Last step in a chain of PRs (#26836, #28894, #28987, #29403).

#### Detailed Description:
The current wallet migration process performs only individual db writes. Accessing disk to
delete all legacy records, clone and clean each address book entry for every created wallet,
create each new descriptor (with their corresponding master key, caches and key pool), and
also clone and delete each transaction that requires to be transferred to a different wallet.

This work consolidates all individual disk writes into two batch operations. One for the descriptors
creation from the legacy data and a second one for the execution of the migration process itself.
Efficiently dumping all the information to disk at once atomically at the end of each process.

This represent a speed up and also a consistency improvement. During migration, we either
want to succeed or fail. No other outcomes should be accepted. We should never leave a
partially migrated wallet on disk and request the user to manually restore the previous wallet from
a backup (at least not if we can avoid it).

Since the speedup depends on the storage device, benchmark results can vary significantly.
Locally, I have seen a 15% speedup on a USB 3.2 pendrive.

#### Note for Testers:
The first commit introduces a benchmark for the migration process. This one can be
cherry-picked on top of master to compare results pre and post changes.

Please note that the benchmark setup may take some time (~70 seconds here) due to the absence
of a batching mechanism for the address generation process (`GetNewDestination()` calls).

furszy committed:
wallet: decouple default descriptors creation from external signer setup

This will be useful in the following-up commit to batch the entire
wallet migration process.

furszy committed:
wallet: introduce 'SetWalletFlagWithDB'

furszy committed:
bench: add coverage for wallet migration process

furszy committed:
wallet: batch MigrateToDescriptor() db transactions

Grouping all db writes into a single atomic write operation.
Speeding up the flow and preventing inconsistent states.

pablomartin4btc reviewed (commented):
Concept ACK.

I'll also perform the benchmark as instructed clearly in the description and will review the code later.

theStack commented:
Concept ACK

furszy commented:
Rebased after #28894 merge, next in the line is #28987.

furszy reviewed (commented):
Rebased, conflicts solved. Next PR in the line is #28987.

furszy commented:
Decoupled the address book cloning related commits to #26836.

josibake commented:
Concept ACK

Starting to review, if you get a chance can you rebase this so that it's only the relevant outstanding commits?

furszy committed:
wallet: provide WalletBatch to 'DeleteRecords'

So it can be used within an external db txn context.

furszy committed:
wallet: refactor ApplyMigrationData to return util::Result<void>

furszy committed:
wallet: migration, consolidate main wallet db writes

Perform a single db write operation for the entire
migration procedure.

furszy reviewed (commented):
> Starting to review, if you get a chance can you rebase this so that it's only the relevant outstanding commits?

Done. Rebased and reworked part of it.
I still need to accommodate some code and improve the commit descriptions. But, at least the changes are reviewable now.

furszy commented on src/bench/wallet_migration.cpp:
```diff
@@ -0,0 +1,87 @@
+
+    int wallet_num = 0;
+    bench.epochs(NUM_WALLETS).run([&] {
+        std::string filename = strprintf("legacy_%d", wallet_num);
```
> I don't like the c_str below. Also, it doesn't compile on Windows.

yeah, me neither.. I just did this 5 months ago. Fixing..

josibake commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1811,8 +1817,8 @@ std::optional<MigrationData> LegacyScriptPubKeyMan::MigrateToDescriptor()
         auto desc_spk_man = std::unique_ptr<DescriptorScriptPubKeyMan>(new DescriptorScriptPubKeyMan(m_storage, w_desc, m_keypool_size));
-        desc_spk_man->AddDescriptorKey(key, key.GetPubKey());
-        desc_spk_man->TopUp();
+        WITH_LOCK(desc_spk_man->cs_desc_man, desc_spk_man->AddDescriptorKeyWithDB(batch, key, key.GetPubKey()));
```
Also, in `mapCryptedKey`, the `CPubKey` is already there, so you could pull it out and avoid recalculating it: https://github.com/bitcoin/bitcoin/pull/28574/commits/5cc89e5eba7330ea3ab64ca66e364170dcd0b4f7#diff-55ce403d9c3440b4a38a261b7d452f84dd9a1d09b2a5ef1b4d3ac0c3567b9106R1762

But I'm not sure that's worth the extra effort considering it would only make a difference if a majority of the keys come from `mapCryptedKey` and it seems that this isn't making much of an improvement anyways.

josibake commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1923,9 +1929,9 @@ std::optional<MigrationData> LegacyScriptPubKeyMan::MigrateToDescriptor()
                     continue;
                 }
-                desc_spk_man->AddDescriptorKey(key, key.GetPubKey());
+                WITH_LOCK(desc_spk_man->cs_desc_man, desc_spk_man->AddDescriptorKeyWithDB(batch, key, key.GetPubKey()));
```
in "wallet: batch MigrateToDescriptor() db transactions" (https://github.com/bitcoin/bitcoin/pull/28574/commits/5cc89e5eba7330ea3ab64ca66e364170dcd0b4f7):

Could get rid of the `key.GetPubKey()` call here and completely get rid of the `privkeyid` set above:

https://github.com/bitcoin/bitcoin/pull/28574/commits/5cc89e5eba7330ea3ab64ca66e364170dcd0b4f7#diff-55ce403d9c3440b4a38a261b7d452f84dd9a1d09b2a5ef1b4d3ac0c3567b9106R1907-R1910

since `keys.origins` is a pair of the `keyid` and `CPubKey`. Running the bench with with this suggestion seemed to help:

```
# after
|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|    6,395,421,573.00 |                0.16 |    0.0% |32,790,285,573.00 |11,895,327,005.00 |  2.757 |3,044,102,935.00 |    0.2% |      6.40 | `WalletMigration`
```

6.94 -> 6.84 -> 6.4

josibake replied:
Makes sense! I misread the title as meaning this PR is both optimizing the process and batching db transactions, but its clear from the PR description that this is focused on just the batching logic.

josibake reviewed (commented):
Looking good! I noticed a lot of `key.GetPubKey()` that could be removed, left some suggestions and verified locally that they improved the bench. I guess one conceptual question is if we are doing the `GetPubKey()` calls to regenerate the pubkeys in case we don't trust the data in the legacy wallet?

furszy reviewed (commented):
Thanks for the review josi!

josibake commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -1782,6 +1782,12 @@ std::optional<MigrationData> LegacyScriptPubKeyMan::MigrateToDescriptor()
 
+    WalletBatch batch(m_storage.GetDatabase());
+    if (!batch.TxnBegin()) {
+        LogPrintf("Error during descriptors migration, cannot initialize db transaction\n");
```
I see, maybe "Error generating descriptors for migration, cannot initialize db transaction"? Also not a big deal, its just he current message is confusing to me in that we aren't migrating descriptors, but taking existing keys, creating descriptors and then saving them in a new wallet.

furszy replied:
Sure. Done as suggested.

josibake commented:
ACK https://github.com/bitcoin/bitcoin/pull/28574/commits/868e2f64a9d7de3a980e64fd7acd7323214d3765

Code reviewed and verified the bench results. Thanks for taking the suggestions!

achow101 commented on src/wallet/scriptpubkeyman.cpp:
```diff
@@ -2009,9 +2009,15 @@ std::optional<MigrationData> LegacyScriptPubKeyMan::MigrateToDescriptor()
 }
 
 bool LegacyScriptPubKeyMan::DeleteRecords()
+{
+    return RunWithinTxn(m_storage.GetDatabase(), /*process_desc=*/"delete legacy records", [&](WalletBatch& batch){
+        return DeleteRecords(batch);
+    });
+}
```
In cd694033173ec7de2a16187432a3d7aa33ff62ca "wallet: introduce 'DeleteRecords(batch)' to be used within an external db txn"

After this PR, this `DeleteRecords()` overload is only used by the tests, so I think it can just be removed (and perhaps never introduced).

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -4280,11 +4266,16 @@ bool DoMigration(CWallet& wallet, WalletContext& context, bilingual_str& error,
+            return false;
+        }
+        // Process succeed, connect the SPKM signals
+        wallet.ConnectScriptPubKeyManNotifiers();
```
In 868e2f64a9d7de3a980e64fd7acd7323214d3765 "wallet: ApplyMigrationData, group all db txs into a single atomic operation"

I don't think it's necessary to connect the signals anymore since the local wallet exists outside of any contexts. Although removing this should be its own commit.

If you choose to not remove this, for completeness, this should also do `NotifyCanGetAddressesChanged` to match the previous behavior.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -2334,12 +2334,22 @@ DBErrors CWallet::LoadWallet()
+    return {};
+}
+
+util::Result<std::vector<CWallet::MapWalletIt>> CWallet::RemoveTxs(WalletBatch& batch, std::vector<uint256>& txs_to_remove) {
```
In 4b3ec698a228fc531cca946f9a572e1dbdc4b0d5 "wallet: prepare RemoveTxs to be used within a broader db txn procedure"

I don't really like this refactor. This means that any callers of this particular overload will have to remember to do `ClearTxs`, whereas callers of the other one don't. This seems like something that could result in programming mistakes in the future. Generally our overloads should do basically the same things, just with the option of providing different arguments. Not clearing the txs from `mapWallet` in one overload is a pretty big difference that may not be easily noticed.

I think it would actually be a bit cleaner and better to have `RemoveTxs` detect if a database transaction has been started already and just choose to not begin a transaction if that's the case.

furszy replied:
Ended up introducing a mechanism to listen for the db txn outcomes. So internal procedures can register callbacks to update the in-memory state only when the data is successfully committed to disk.

furszy committed:
wallet: remove post-migration signals connection

The wallet is isolated during migration and reloaded at the end
of the process. There is no benefit on connecting the signals
few lines before unloading the wallet.

furszy committed:
wallet: provide WalletBatch to 'RemoveTxs'

Preparing it to be used within a broader db txn procedure.

furszy committed:
wallet: introduce active db txn listeners

Useful to ensure that the in-memory state is updated only
after successfully committing the data to disk.

furszy reviewed (commented):
Updated per feedback. Thanks achow.

furszy commented:
rebased after #29130 merge.

furszy reviewed (commented):
Rebased due a hidden conflict with #26606.

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -3696,25 +3696,17 @@ DescriptorScriptPubKeyMan& CWallet::SetupDescriptorScriptPubKeyMan(WalletBatch&
 }
 
-void CWallet::SetupDescriptorScriptPubKeyMans(const CExtKey& master_key)
+void CWallet::SetupDescriptorScriptPubKeyMans(WalletBatch& batch, const CExtKey& master_key)
```
In 4e72821e1961b58559f49a4746bc8d15d116f6b7 "wallet: ApplyMigrationData, group all db txs into a single atomic operation"

This could be its own commit.

furszy replied:
Done

furszy replied:
Done

achow101 reviewed (commented):
Concept ACK

I'm not a huge fan of all the boilerplate and wrappers that this ends up adding, but I also don't see a better way to do this.

In `ApplyMigrationData`, it looks like we're creating batchs for the watchonly and solvable wallets multiple times. Those could be consolidated in this PR as well.

furszy committed:
wallet: provide WalletBatch to 'SetupDescriptorScriptPubKeyMans'

So it can be used within an external db txn context.

furszy committed:
wallet: migration, consolidate external wallets db writes

Perform a single db write operation for each external wallet
(watch-only and solvables) for the entire migration procedure.

furszy reviewed (commented):
> In ApplyMigrationData, it looks like we're creating batchs for the watchonly and solvable wallets multiple times. Those could be consolidated in this PR as well.

Done. Thanks.

achow101 commented:
ACK ad1e3620af88f73b869b4a2dbb74e03f5cd93517

HerWayBit commented on src/bench/wallet_migration.cpp:
```diff
@@ -0,0 +1,83 @@
+#include <bench/bench.h>
+#include <interfaces/chain.h>
+#include <node/context.h>
+#include <test/util/mining.h>
```
https://cash.app/$SelfMadeLady270

hebasto commented on src/Makefile.bench.include:
```diff
@@ -95,6 +95,7 @@ bench_bench_bitcoin_SOURCES += bench/wallet_create.cpp
 bench_bench_bitcoin_SOURCES += bench/wallet_loading.cpp
 bench_bench_bitcoin_SOURCES += bench/wallet_create_tx.cpp
 bench_bench_bitcoin_SOURCES += bench/wallet_ismine.cpp
+bench_bench_bitcoin_SOURCES += bench/wallet_migration.cpp
```
Changes to the Autotools-based build system might be dropped altogether. Otherwise, a conflict is unavoidable once https://github.com/bitcoin/bitcoin/pull/30664 is merged, which is expected to happen tomorrow.

furszy replied:
pushed.

theStack reviewed (commented):
I'm not seeing any performance improvements locally with a non-debug build running on an SSD (benchmark shows ~13s both with and without the batching), but still think this PR is important for consistency reasons, i.e. to avoid the new wallet be left in a "half-migrated" state in case anything goes wrong. Changes look good to me at first glance (left only some nits), will do a deeper review round tomorrow.

Might be worth to update the PR description with benchmark results from release builds, as only those are relevant for end users and the speed-up could vary significantly?

furszy commented on src/bench/wallet_migration.cpp:
```diff
@@ -0,0 +1,83 @@
+#include <optional>
+
+#ifdef USE_BDB // only enable benchmark when bdb and sqlite are enabled
+#ifdef USE_SQLITE
```
done

furszy commented on src/bench/wallet_migration.cpp:
```diff
@@ -0,0 +1,83 @@
+    std::vector<CScript> scripts_watch_only;
+    for (int w = 0; w < NUM_WATCH_ONLY_ADDR; ++w) {
+        CKey key;
+        key.MakeNewKey(true);
```
done

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -1701,10 +1701,16 @@ bool CWallet::CanGetAddresses(bool internal) const
+    return SetWalletFlagWithDB(batch, flags);
+}
+
+void CWallet::SetWalletFlagWithDB(WalletBatch& batch, uint64_t flags)
```
I think this was part of the first changes I did, I do like more the overloaded function approach as is less verbose without loosing readability. But later on I saw that we are using the `WithDB` suffix and changed the rest of the commits.
Will change `DeleteRecords` to use it. Thanks.

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -3746,21 +3746,27 @@ void CWallet::SetupDescriptorScriptPubKeyMans(const CExtKey& master_key)
 
+void CWallet::SetupOwnDescriptorScriptPubKeyMans()
+{
+    assert(!IsWalletFlagSet(WALLET_FLAG_EXTERNAL_SIGNER));
```
sure. done

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -4104,12 +4099,13 @@ util::Result<void> CWallet::ApplyMigrationData(MigrationData& data)
         } else {
             // Setup with a new seed if we don't.
-            SetupOwnDescriptorScriptPubKeyMans();
+            SetupOwnDescriptorScriptPubKeyMans(local_wallet_batch);
```
> in [c436b51](https://github.com/bitcoin/bitcoin/commit/c436b516450b35bb01393559d935e1b807e5a835): shouldn't those calls also happen as a batch transaction, i.e. wrapped around a TxnBegin/TxnCommit pair (at least in the context of this commit)?

yeah, they should, but that would make this commit and the last one more verbose without much gain. Maybe it's better to leave it as is for simplicity?

furszy commented:
Updated per feedback, thanks @theStack!

> I'm not seeing any performance improvements locally with a non-debug build running on an SSD (benchmark shows ~13s both with and without the batching), but still think this PR is important for consistency reasons, i.e. to avoid the new wallet be left in a "half-migrated" state in case anything goes wrong. Changes look good to me at first glance (left only some nits), will do a deeper review round tomorrow.
>
> Might be worth to update the PR description with benchmark results from release builds, as only those are relevant for end users and the speed-up could vary significantly?

Yeah, after all the other PRs got merged, this one is important for people running on HDDs. See https://github.com/bitcoin/bitcoin/pull/28037#issuecomment-1952211224, which shared a pretty nice time improvement. Going from a 30 minutes migration to one that took only 80 seconds.
Still, that was a long time ago. Will give it a new run on an HDD and update the PR description.

theStack reviewed (approved):
Code-review ACK a4e1ba722be5dbd71a3d21e5da82aee5409d22bd

Thanks for following up! 

> Yeah, now that all the other PRs got merged, things changed a bit. Aside from the consistency reasons, this PR is important for people running on HDDs. See https://github.com/bitcoin/bitcoin/pull/28037#issuecomment-1952211224, which shared a pretty nice time improvement. Going from a 30 minutes migration to one that took only 80 seconds.
Still, that was a long time ago. Will give it a new run on an HDD and update the PR description.

That sounds pretty impressive (>20x?). Would be very interesting indeed to re-benchmark with this branch again (cc @willcl-ark), though as said above the changes also make sense to me if the performance improvement is not that significant.

furszy commented:
PR description updated. Created #31000 so we can all run the benchmark on an external HDD in a more friendly manner.
Have seen a 15% speedup locally on a USB 3.2 pendrive.

pablomartin4btc reviewed (commented):
ACK a4e1ba722be5dbd71a3d21e5da82aee5409d22bd

<details>
<summary>I've ran the benchmarks locally on SSD and on an external USB 3.2 pendrive and got an improvement of ~46% (i7-1260P, 32GB RAM, Ubuntu 22.04 - using <code>pyperf system tune</code> to get stable results).</summary>

- This PR:

```
./build/src/bench/bench_bitcoin -filter=WalletMigration

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|    5,322,062,123.00 |                0.19 |    0.0% |      5.32 | `WalletMigration`
```

- `master`

```
./build/src/bench/bench_bitcoin -filter=WalletMigration

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|   11,362,482,640.00 |                0.09 |    0.0% |     11.36 | `WalletMigration`

```
</details>

furszy commented:
@josibake @achow101 friendly ping. It seems we are quite close here.

furszy commented:
Sad rebase due to tiny conflict with #30937. Ready to go.

achow101 commented:
ACK c98fc36d094a08d44f3c95431db2c5f34a96cc73

theStack reviewed (approved):
re-ACK c98fc36d094a08d44f3c95431db2c5f34a96cc73
(as per `$ git range-diff a4e1ba7...c98fc36`)

pablomartin4btc reviewed (approved):
re-ACK c98fc36d094a08d44f3c95431db2c5f34a96cc73

(no diff since my previous review, just rebased)
